### PR TITLE
Feat/ Attachment Controller

### PR DESCRIPTION
**Pull request by nitro-marky**
Attachment controller without validation. 

Thanks to @n3op2

### CODE REVIEW

File: app/api-v1/controllers/Attachment/index.js [hunk 1]
```diff
@@ -1,5 +1,73 @@
...
   },
+  get: async (req) => {
+    const { id } = req.params
+    if (!id) throw new BadRequestError('missing params')
+
+    const [attachment] = await db.getAttachment(req.params.id)
+    if (!attachment) throw new NotFoundError('Attachment Not Found')
+    const orderedAccept = parseAccept(req.headers.accept)
+
+    for (const mimeType of orderedAccept) {
```
Comment (n3op2) [inline app/api-v1/controllers/Attachment/index.js:21]: This is naughty implementation, but does what it should. Will add to the backlog for DA PHUTURE

**Review by n3op2 — DISMISSED** [review]

File: app/api-v1/controllers/Attachment/index.js [hunk 2]
```diff
@@ -1,5 +1,73 @@
...
+        return {
+          status: 200,
+          headers: {
+            immutable: true,
+            maxAge: 365 * 24 * 60 * 60 * 1000,
+            'content-disposition': `attachment; filename="${attachment.filename}"`,
+            'access-control-expose-headers': 'content-disposition',
+            'content-type': 'application/octet-stream',
+          },
+          response: attachment.binary_blob.toString(),
```
Comment (mattdean-digicatapult) [inline app/api-v1/controllers/Attachment/index.js:43]: Pretty sure this should not be doing a `toString()`. I commented before how that likely results in an invalid utf8 conversaion

File: app/api-v1/routes/attachment/{id}.js [hunk 3]
```diff
@@ -1,58 +1,11 @@
...
-            })
-            res.send(response.binary_blob)
-            return
-          }
-          throw new NotAcceptableError({ message: 'Client file request not supported' })
-        }
-      }
+    GET: async (req, res) => {
+      const { status, response } = await attachmentController.get(req)
+      res.status(status).send(response)
```
Comment (mattdean-digicatapult) [inline app/api-v1/routes/attachment/{id}.js:8]: Headers handling?

File: app/api-v1/routes/attachment/{id}.js [hunk 4]
```diff
@@ -94,6 +47,9 @@ module.exports = function (attachmentService) {
                   type: 'array',
                   items: {},
                 },
+                {
+                  type: 'string',
+                },
```
Comment (mattdean-digicatapult) [inline app/api-v1/routes/attachment/{id}.js:52]: Why have we added this?

File: app/utils/routeResponseValidator.js [hunk 5]
```diff
@@ -35,15 +35,18 @@ const buildValidatedJsonHandler = (controller, apiDoc) => {
     responses,
     components: commonApiDoc.components,
   })
+
   const handler = async function (req, res) {
-    const { status, response } = await controller(req)
+    const { status, response, headers } = await controller(req)
```
Comment (mattdean-digicatapult) [inline app/utils/routeResponseValidator.js:40]: I don't believe any code is using these enhancements now. Is there a reason to keep the additional complexity introduced here?

File: test/integration/attachment.test.js [hunk 6]
```diff
@@ -103,8 +103,8 @@ describeAuthOnly('attachments - authenticated', function () {
   })
 
   it('should return 406 when requesting json from the octet route', async function () {
-    const attachment = '00000000-0000-1000-8000-000000000001'
-    const response = await getAttachmentRouteJSON(attachment, app, authToken)
+    const attachment = '00000000-0000-1000-9000-000000000001'
+    const response = await getAttachmentRouteOctet(attachment, app, authToken)
```
Comment (mattdean-digicatapult) [inline test/integration/attachment.test.js:107]: I don't think this is right? The test is checking what happens is we request JSON from the octet-stream value (which looking at the seed is `00000000-0000-1000-8000-000000000001`)

**Review by mattdean-digicatapult — COMMENTED** [review]
Questions that will likely result in changes

File: app/api-v1/controllers/Attachment/index.js (hunk 1 above)
Comment (mattdean-digicatapult) [reply app/api-v1/controllers/Attachment/index.js:21]: why is this naughty?

File: app/utils/routeResponseValidator.js (hunk 5 above)
Comment (n3op2) [reply app/utils/routeResponseValidator.js:40]: yes, for handling dynamic headers, default is application/json so it does not break anything, as third argument is optional

File: test/integration/attachment.test.js (hunk 6 above)
Comment (n3op2) [reply test/integration/attachment.test.js:107]: Mmm... I think you are correct. @nitro-marky any ideas?

File: app/api-v1/routes/attachment/{id}.js (hunk 3 above)
Comment (n3op2) [reply app/api-v1/routes/attachment/{id}.js:8]: Sorry, I missed that. Was mainly focused on getting this working. @nitro-marky leave it to me I will update now. Thanks for spotting, clearly test coverage is not complete.

File: app/api-v1/routes/attachment/{id}.js (hunk 4 above)
Comment (n3op2) [reply app/api-v1/routes/attachment/{id}.js:52]: accident, leftover from buildValiudator, it has been removed

File: test/integration/attachment.test.js (hunk 6 above)
Comment (mattdean-digicatapult) [reply test/integration/attachment.test.js:107]: right which is JSON. This route should request JSON but from an octet stream only source. That's `00000000-0000-1000-8000-000000000001`

File: app/api-v1/controllers/Attachment/index.js (hunk 2 above)
Comment (mattdean-digicatapult) [reply app/api-v1/controllers/Attachment/index.js:43]: why has this been resolved? I don't believe this was updated?

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]

File: app/api-v1/controllers/Attachment/index.js (hunk 2 above)
Comment (n3op2) [reply app/api-v1/controllers/Attachment/index.js]: sorry my changes have gone to wrong branch

File: app/api-v1/controllers/Attachment/index.js (hunk 1 above)
Comment (n3op2) [reply app/api-v1/controllers/Attachment/index.js]: There a couple of reasons. I thnk we are under utilising helper method where all the heavy lifting if happening.

Well, it's just to me looks naughty or something might not do, but again I might not know some underlying  reasons

File: test/integration/attachment.test.js [hunk 7]
```diff
@@ -102,9 +102,9 @@ describeAuthOnly('attachments - authenticated', function () {
     expect(response.status).to.equal(200)
   })
 
-  it('should return 406 when requesting json from the octet route', async function () {
-    const attachment = '00000000-0000-1000-8000-000000000001'
-    const response = await getAttachmentRouteJSON(attachment, app, authToken)
+  it('should return 406 when filename is .json but accept header is octet', async function () {
+    const attachment = '00000000-0000-1000-9000-000000000001'
+    const response = await getAttachmentRouteOctet(attachment, app, authToken)
```
[the remark this answers is not in the document]
Comment (n3op2) [reply test/integration/attachment.test.js]: This has been updated

**Review by n3op2 — DISMISSED** [review]

**Review by n3op2 — DISMISSED** [review]

**Review by n3op2 — DISMISSED** [review]

File: app/api-v1/controllers/Attachment/__tests__/attachments.test.js [hunk 8]
```diff
@@ -1,9 +1,115 @@
...
+
+const getAttachment = async (req) => {
+  try {
+    return await attachmentController.get(req)
+  } catch (err) {
+    return err
+  }
+}
+
+describe('Attachment controller', () => {
```
Comment (mattdean-digicatapult) [inline app/api-v1/controllers/Attachment/__tests__/attachments.test.js:17]: I think we could have some better coverage here. We're not really coverring passing multiple accept header entries and the prioritisation that comes of that. That said if both @n3op2 and @nitro-marky are happy we can leave this for another day.

File: app/api-v1/controllers/Attachment/__tests__/attachments.test.js [hunk 9]
```diff
@@ -1,9 +1,115 @@
...
+      beforeEach(async () => {
+        response = await getAttachment({ params: {} })
+      })
+
+      it('throws a validation error', async () => {
+        expect(response).to.be.an.instanceOf(BadRequestError)
+        expect(response.message).to.be.equal('Bad Request: missing params')
+      })
+
+      it('does not perform any database calls and does not create transaction', () => {
```
Comment (mattdean-digicatapult) [inline app/api-v1/controllers/Attachment/__tests__/attachments.test.js:44]: copy paste? I don't think Attachment GET ever creates a transaction

**Review by mattdean-digicatapult — DISMISSED** [review]
Approved. I think we could have some better test coverage. I don't think the refactored create transaction has been given any tests for example. Also spotted at least one errant description in a test. Not enough for me to prevent this merging

File: app/api-v1/controllers/Attachment/index.js [hunk 10]
```diff
@@ -1,5 +1,75 @@
...
+      throw new NotAcceptableError({ message: 'Client file request not supported' })
+    }
+    for (const mimeType of orderedAccept) {
+      if (mimeType === 'application/octet-stream' || mimeType === 'application/*' || mimeType === '*/*') {
+        return returnOctet(attachment)
+      }
+    }
+    throw new NotAcceptableError({ message: 'Client file request not supported' })
+  },
+  create: async (req) => {
```
[the remark this answers is not in the document]
Comment (nitro-marky) [reply app/api-v1/controllers/Attachment/index.js:52]: Yeah, a separate PR makes sense

**Review by n3op2 — APPROVED** [review]

**Pull request merged**